go/libraries/doltcore/sqle/index/z_encoding.go

```diff
@@ -184,28 +179,118 @@ func ZCell(v types.GeometryValue) val.Cell {
 	return ZMask(level, zMin)
 }
 
-// ZAddr converts the GeometryValue into a key: (level, min_z_val)
-func ZAddr(v types.GeometryValue) val.Cell {
-	bbox := spatial.FindBBox(v)
-	zMin := ZValue(types.Point{X: bbox[0], Y: bbox[1]})
-	zMax := ZValue(types.Point{X: bbox[2], Y: bbox[3]})
-	addr := val.Cell{}
-	binary.BigEndian.PutUint64(addr[1:], zMin[0])
-	binary.BigEndian.PutUint64(addr[9:], zMin[1])
-	if res := zMin[0] ^ zMax[0]; res != 0 {
-		addr[0] = byte(64 - bits.LeadingZeros64(res)/2)
+// ZRange is a pair of two ZVals
+// ZRange[0] is the lower bound (z-min)
+// ZRange[1] is the upper bound (z-max)
+type ZRange = [2]ZVal
+
+// mergeZRanges combines the z-ranges in acc with zRange by either
+// 1. combining the last ZRange in acc with zRange if the ranges are next to each other or
+// 2. appending zRange to acc
+func mergeZRanges(acc []ZRange, zRange ZRange) []ZRange {
+	n := len(acc) - 1
+	if n >= 0 && acc[n][1][0] == zRange[0][0] && zRange[0][1]-acc[n][1][1] == 1 {
+		acc[n][1] = zRange[1]
+		return acc
+	}
+	return append(acc, zRange)
+}
+
+// zRangeSize retrieves the approximate size of the zRange
+// it only takes the top 64 bits of the difference
+// it accepts and returns a shift-amount so that comparison between two zRangeSizes are consistent
+func zRangeSize(zRange ZRange, shamt int) (uint64, int) {
+	zVal := ZVal{}
+	zVal[0] = zRange[1][0] - zRange[0][0]
+	if zRange[1][1] < zRange[0][1] {
+		zVal[0] -= 1
+		zVal[1] = ^zRange[1][1] - zRange[0][1]
 	} else {
-		addr[0] = byte(32 - bits.LeadingZeros64(zMin[1]^zMax[1])/2)
+		zVal[1] = zRange[1][1] - zRange[0][1]
 	}
-	return addr
+	if shamt == -1 {
+		shamt = bits.LeadingZeros64(zVal[0])
+	}
+	zVal[0] = zVal[0] << shamt
+	zVal[1] = zVal[1] >> (64 - shamt)
+	return zVal[0] | zVal[1], shamt
 }
 
-// ZAddrSort converts the GeometryValue into a key: (min_z_val, level)
-// Note: there is an inefficiency here where small polygons may be placed into a level that's significantly larger
-func ZAddrSort(geoms []types.GeometryValue) []types.GeometryValue {
-	sort.Slice(geoms, func(i, j int) bool {
-		zi, zj := ZAddr(geoms[i]), ZAddr(geoms[j])
-		return bytes.Compare(zi[:], zj[:]) < 0
-	})
-	return geoms
+// Thresholds to stop splitting ZRanges
+const cutThresh = 0.02
+const depthThresh = 4
+
+// shouldCut checks if the size of the removed ZRange divided by the size of the whole ZRange is smaller than cutThresh, stop
+func shouldCut(cutRange ZRange, size float64, shamt int) bool {
+	cut, _ := zRangeSize(cutRange, shamt)
+	return (float64(cut) / size) >= cutThresh
+}
+
+// isContinuous checks if the provided zRange is entirely within the bounding box
+func isContinuous(zl, zh uint64, prefixLength int) bool {
+	mask := uint64(math.MaxUint64 >> prefixLength)
+	return (zl&mask) == 0 && (zh&mask) == mask
+}
+
+// splitZRanges is a helper function to SplitZRanges
+func splitZRanges(zRange ZRange, zSize float64, zShamt, depth int, acc []ZRange) []ZRange {
```

Comment:
This function is pretty impossible to read. I think most all devs will find it opaque if they need to modify it. At minimum you should replace all the literals with constants and try to describe their purpose